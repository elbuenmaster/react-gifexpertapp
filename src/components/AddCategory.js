import React, { useState } from "react";
import Proptypes from 'prop-types'
export const AddCategory = ({ setCategorias }) => {
  
  const [inputValue, setInputValue] = useState('');

  const handleInputValue = (e) => {
    setInputValue(e.target.value);
    console.log('se cambio el valor de la cajita','{',e.target.value,'}');
  };

  const handleSubmit = (e) => {
              e.preventDefault();
              console.log("se dispara Submit");
              if (inputValue.trim().length > 2) {
                // setCategorias: se le agrega, usando una funcion CALLBACK que lea el valor del State y le aÃ±ada el nuevo valor
                setCategorias((cats) => [ inputValue,...cats]);
                setInputValue('');
                console.log("Submit exitoso");
                return true;
              }
              console.log("se cancela Submit");
  };

  return (
    <form onSubmit={handleSubmit}>
      <p>{inputValue}</p>
      <input type="text" value={inputValue} onChange={handleInputValue}></input>
    </form>
  );
};

AddCategory.propTypes={
    setCategorias: Proptypes.func.isRequired
}