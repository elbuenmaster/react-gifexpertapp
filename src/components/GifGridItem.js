import React from 'react'
import PropTypes from 'prop-types'
export const GifGridItem = ({id,title,url}) => {
    console.log(id)
    return (
        <div className="card animate__fadeIn">
        
           <img src={url} alt={title}></img> 
           <p>{title}</p>       
       
        </div>
    )
}

GifGridItem.propTypes={
    id:PropTypes.number.isRequired,
    title:PropTypes.string.isRequired,
    url:PropTypes.string.isRequired,
}
export default GifGridItem;