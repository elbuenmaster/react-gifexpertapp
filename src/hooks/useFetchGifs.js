import { useEffect, useState } from "react";
import { getGifs } from "../helpers/getGifs";
export const useFetchGifs = (url) => {
  const [state, setstate] = useState({
    data: [],
    loading: true,
  });

  useEffect(() => {
    getGifs(url).then((imgs) => {
      console.log(imgs);
      setstate({
        data: imgs,
        loading: false,
      });
    });
  }, [url]);

  return state; // retorna el objeto completo : array + boolean
};
