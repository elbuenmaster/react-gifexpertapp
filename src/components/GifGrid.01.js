import React, { useState, useEffect } from "react";
import { GifGridItem } from "../components/GifGridItem";
import {getGifs} from "../helpers/getGifs";

export const GifGrid = ({ category }) => {
  const apiKey = "eBHH7BLGeQviyhM8FyoIaVveWFeN72yl";
  const url = `https://api.giphy.com/v1/gifs/search?q=${encodeURI(category)}&limit=10&api_key=${apiKey}`;

  const [images, setImages] = useState([]);

  

  useEffect(() => {
    getGifs(url)
    .then(imgs=>setImages(imgs));
  }, [url]);

  return (
    <>  
    <h3>{category}</h3>
    <div className='card-grid'>
    
      
        {images.map((imagen) => (
          <GifGridItem 
          key={imagen.id}
          {...imagen} />
        ))}
     
    </div>
   </> 
  );
};
