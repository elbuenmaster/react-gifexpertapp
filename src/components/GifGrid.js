import React, { useState, useEffect } from "react";
import { GifGridItem } from "../components/GifGridItem";
import { useFetchGifs } from "../hooks/useFetchGifs";
import PropTypes from 'prop-types';

export const GifGrid = ({ category }) => {
 
  /* const [images, setImages] = useState([]);
  useEffect(() => {
    getGifs(url)
    .then(imgs=>setImages(imgs));
  }, [url]);
 */

  // al cargar la app por primera vez,data est√° vacio y loading esta en false
  const {data:images,loading}= useFetchGifs(category);
 
  return (
    <>  
    <h3>{category}</h3>
    {loading && <p>Cargando</p>}
   {   <div className='card-grid'>
    
      
        {images.map((imagen) => (
          <GifGridItem 
          key={imagen.id}
          {...imagen} />
        ))}
     
    </div> }
   </> 
  );
};

GifGrid.ProTypes={
  category:PropTypes.string.isRequired
}