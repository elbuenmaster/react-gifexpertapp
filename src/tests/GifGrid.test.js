import React from 'react';
import {render} from '@testing-library/react';
import  '@testing-library/jest-dom';
import {shallow} from  'enzyme';
import {GifGrid} from '../components/GifGrid';
import {useFetchGifs} from '../hooks/useFetchGifs' 
// esto es un Mock CustomHook
// mock es para simular comportamientos
jest.mock('../hooks/useFetchGifs');

describe('testing GridItem', () => {
    let wrapper;
    let categoria='Fury Road';
    

    test('correcto snapshot del componente',()=>{

        // mock es para simular comportamientos
        // en este caso, el hook intervenido useFetchGifs
        // retorna una array vacio  y el boleano en false.
        // tal como debe suceder a cargar por primera vez el componente.
         useFetchGifs.mockReturnValue({
            data:[],
            loading:true,
        });  
        let  wrapper = shallow(<GifGrid category={categoria}/>);
        expect(wrapper).toMatchSnapshot();
    })




    test('debe mostrar items cuando se cargan imagenes con mi hook - useEffect',()=>{
            const gifs = [{
                id:1,
                url:'https://www.1.cl/hola.jpg',
                title:'mi foto'
            },
            {
                id:2,
                url:'https://www.1.cl/hol222a.jpg',
                title:'mi foto 2'
            },
          ]
            useFetchGifs.mockReturnValue({
                data:gifs,
                loading:false,
            });  
            let  wrapper = shallow(<GifGrid category={categoria}/>);
            expect(wrapper).toMatchSnapshot();

            expect(wrapper.find('p').exists()).toBe(false);

            // busquemos si  contiene  el componente 'GifGridItem' 
            expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
           
    })

})
