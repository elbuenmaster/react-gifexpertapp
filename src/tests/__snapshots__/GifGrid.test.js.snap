// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing GridItem correcto snapshot del componente 1`] = `
<Fragment>
  <h3>
    Fury Road
  </h3>
  <p>
    Cargando
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`testing GridItem debe mostrar items cuando se cargan imagenes con mi hook - useEffect 1`] = `
<Fragment>
  <h3>
    Fury Road
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id={1}
      key="1"
      title="mi foto"
      url="https://www.1.cl/hola.jpg"
    />
    <GifGridItem
      id={2}
      key="2"
      title="mi foto 2"
      url="https://www.1.cl/hol222a.jpg"
    />
  </div>
</Fragment>
`;
